### PR TITLE
Preview: MQTT parser - v1

### PR DESCRIPTION
**Pull request by satta**
- [X] I have read the contributing guide lines at https://redmine.openinfosecfoundation.org/projects/suricata/wiki/Contributing
- [X] I have signed the Open Information Security Foundation contribution agreement at https://suricata-ids.org/about/contribution-agreement/
- [X] I have updated the user guide (in doc/userguide/) to reflect the changes made (if applicable)

Link to [redmine](https://redmine.openinfosecfoundation.org/projects/suricata/issues) ticket:  [#3549](https://redmine.openinfosecfoundation.org/issues/3549)

This is a preview of an almost complete (in my opinion) MQTT (https://mqtt.org) parser for Suricata 6. Implemented are the following features:

- Support for MQTT versions 3.1, 3.1.1 and 5.0 (from legacy to current)
- EVE-JSON logging for all content, including all MQTT 5 properties (see [doc here](https://github.com/satta/suricata/blob/mqtt/doc/userguide/output/eve/eve-json-format.rst#event-type-mqtt) for examples)
- New keywords for matching on fixed and variable header fields, flags, and payloads (see [doc here](https://github.com/satta/suricata/blob/mqtt/doc/userguide/rules/mqtt-keywords.rst#mqtt-keywords) for descriptions)
    - `mqtt.type`
    - `mqtt.flags`
    - `mqtt.qos`
    - `mqtt.protocol_version`
    - `mqtt.reason_code`
    - `mqtt.connect.clientid`
    - `mqtt.connect.flags`
    - `mqtt.connect.password`
    - `mqtt.connect.username`
    - `mqtt.connect.willmessage`
    - `mqtt.connect.willtopic`
    - `mqtt.connack.return_code`
    - `mqtt.connack.session_present`
    - `mqtt.publish.message`
    - `mqtt.publish.topic`
    - `mqtt.subscribe.topic`
    - `mqtt.unsubscribe.topic`
- MQTT metadata embedded in alerts
- suricata-verify test cases with >240 checks for all protocol versions and message types, in a separate branch of that repo -> https://github.com/OISF/suricata-verify/pull/205

Missing so far:

- Optional transactional grouping (logging request and response(s) for `PUBLISH`, `SUBSCRIBE`, `CONNECT` and `UNSUBSCRIBE` in a single TX) -- this should be simple to add though)
- Use of macros in unit tests

I would be happy to get feedback.

### CODE REVIEW

File: src/detect-mqtt-connack-sessionpresent.c [hunk 1]
```diff
@@ -0,0 +1,328 @@
...
+void MQTTConnackSessionPresentRegisterTests(void);
+void DetectMQTTConnackSessionPresentFree(void *);
+
+static int DetectEngineInspectMQTTConnackSessionPresentGeneric(ThreadVars *tv,
+        DetectEngineCtx *de_ctx, DetectEngineThreadCtx *det_ctx,
+        const Signature *s, const SigMatchData *smd,
+        Flow *f, uint8_t flags, void *alstate,
+        void *txv, uint64_t tx_id);
+
+typedef struct DetectMQTTConnackSessionPresentData_ {
```
Comment (catenacyber) [inline src/detect-mqtt-connack-sessionpresent.c:54]: Why wrap a single bool into a `struct` ?

File: src/detect-mqtt-connack-sessionpresent.c [hunk 2]
```diff
@@ -0,0 +1,328 @@
...
+ *
+ * \param rawstr Pointer to the user provided options
+ *
+ * \retval de pointer to DetectMQTTConnackSessionPresentData on success
+ * \retval NULL on failure
+ */
+static DetectMQTTConnackSessionPresentData *DetectMQTTConnackSessionPresentParse(const char *rawstr)
+{
+    DetectMQTTConnackSessionPresentData *de = NULL;
+#define MAX_SUBSTRINGS 30
```
Comment (catenacyber) [inline src/detect-mqtt-connack-sessionpresent.c:138]: Idea : Could we rather do this parsing in Rust ?

File: src/detect-mqtt-connect-flags.c [hunk 3]
```diff
@@ -0,0 +1,423 @@
...
+#include "detect.h"
+#include "detect-parse.h"
+#include "detect-engine.h"
+#include "detect-engine-content-inspection.h"
+#include "detect-mqtt-connect-flags.h"
+#include "util-unittest.h"
+
+#include "rust-bindings.h"
+
+#define PARSE_REGEX "(?: *,?!?(?:username|password|will|will_retain|clean_session))+"
```
Comment (catenacyber) [inline src/detect-mqtt-connect-flags.c:35]: Same idea : do it in Rust instead of PCRE ?

File: src/detect-mqtt-protocol-version.c [hunk 4]
```diff
@@ -0,0 +1,317 @@
...
+                               void *txv, const Signature *s,
+                               const SigMatchCtx *ctx)
+{
+    const DetectMQTTProtocolVersionData *de = (const DetectMQTTProtocolVersionData *)ctx;
+    uint8_t version;
+
+    if (!de)
+        return 0;
+
+    rs_mqtt_tx_get_protocol_version(state, &version);
```
Comment (catenacyber) [inline src/detect-mqtt-protocol-version.c:118]: Would detect-engine-uint.h be useful here ?
If you want to specify in the signature version>3 or some similar stuff ?

File: rust/src/mqtt/mqtt.rs [hunk 5]
```diff
@@ -0,0 +1,451 @@
...
+#[no_mangle]
+pub extern "C" fn rs_mqtt_probing_parser(
+    _flow: *const Flow,
+    _direction: u8,
+    input: *const u8,
+    input_len: u32,
+    _rdir: *mut u8,
+) -> AppProto {
+    let buf = build_slice!(input, input_len as usize);
+    if parse_message(buf, 3).is_ok() {
```
Comment (catenacyber) [inline rust/src/mqtt/mqtt.rs:241]: Why magic 3 ?

File: rust/src/mqtt/mqtt.rs [hunk 6]
```diff
@@ -0,0 +1,451 @@
...
+        return unsafe { ALPROTO_MQTT };
+    }
+    return ALPROTO_UNKNOWN;
+}
```
Comment (catenacyber) [inline rust/src/mqtt/mqtt.rs:245]: Can we not return ALPROTO_FAILED in some cases ?

File: rust/src/mqtt/mqtt.rs [hunk 7]
```diff
@@ -0,0 +1,451 @@
...
+pub unsafe extern "C" fn rs_mqtt_register_parser() {
+    let default_port = CString::new("[1883]").unwrap();
+    let parser = RustParser {
+        name: PARSER_NAME.as_ptr() as *const std::os::raw::c_char,
+        default_port: default_port.as_ptr(),
+        ipproto: IPPROTO_TCP,
+        probe_ts: Some(rs_mqtt_probing_parser),
+        probe_tc: Some(rs_mqtt_probing_parser),
+        min_depth: 0,
+        max_depth: 16,
```
Comment (catenacyber) [inline rust/src/mqtt/mqtt.rs:413]: Why magic 16 ?

**Comment by norg** [comment]
> What about Suricata-verify tests ?

https://github.com/OISF/suricata-verify/pull/205 :)

**Comment by satta** [comment]
> Here are a few comments.
> This is not complete, but it is better than nothing for now ;-)

True, see below ;)

> Are there no app-layer decoder events for this new protocol ?

Not for now. Most protocol restrictions worthy of events, IMHO, refer to the state of the communication (i.e. `PUBACK`s for messages IDs that were never sent) but at the moment the parser looks at individual messages without context. I could take a look what checks would apply to individual messages, but it wasn't a priority for me right now. (Other protocols also don't have events, I regarded it as a nice-to-have but not essential -- I might be wrong here, never really used events).

> What about Suricata-verify tests ?

They are in a branch, thanks to @norg for posting the link.

File: rust/src/json.rs [hunk 8]
```diff
@@ -127,6 +127,11 @@ impl Json {
             json_array_append_new(self.js, json_string(to_cstring(val.as_bytes()).as_ptr()));
         }
     }
+    pub fn array_append_integer(&self, val: u64) {
```
[the remark this answers is not in the document]
Comment (satta) [reply rust/src/json.rs:130]: True. Will restructure in the next PR.

File: src/detect-mqtt-connack-sessionpresent.c (hunk 1 above)
Comment (satta) [reply src/detect-mqtt-connack-sessionpresent.c:54]: It made handling the boilerplate code easier for that many keywords. But I can polish it a bit and make it more idiomatic.

File: src/detect-mqtt-connect-flags.c (hunk 3 above)
Comment (satta) [reply src/detect-mqtt-connect-flags.c:35]: I just didn't want to experiment here and just stick with the established ways of doing things. I'd say this should probably be in a separate PR?

File: src/detect-mqtt-connack-sessionpresent.c (hunk 2 above)
Comment (satta) [reply src/detect-mqtt-connack-sessionpresent.c:138]: See below.

File: src/detect-mqtt-protocol-version.c (hunk 4 above)
Comment (satta) [reply src/detect-mqtt-protocol-version.c:118]: Makes sense.

File: src/detect-mqtt-connect-flags.c (hunk 3 above)
Comment (catenacyber) [reply src/detect-mqtt-connect-flags.c:35]: @victorjulien thoughts here ?
cf https://github.com/OISF/suricata/pull/4775#discussion_r404354783

**Comment by catenacyber** [comment]
>> Are there no app-layer decoder events for this new protocol ?
>
> Not for now. 

Ok, no worries. I just wanted to make sure you gave a thought about it which seems to be the case.

File: rust/src/mqtt/mqtt.rs (hunk 6 above)
Comment (satta) [reply rust/src/mqtt/mqtt.rs:245]: I think we can. I was under the assumption that `ALPROTO_FAILED` meant something else. I'll change that in v2.

File: rust/src/mqtt/mqtt.rs (hunk 7 above)
Comment (satta) [reply rust/src/mqtt/mqtt.rs:413]: This was just the default from the template, as seen in other implementations as well. I'd be happy to change it, but I couldn't see any negative effects in my cases.

File: rust/src/mqtt/mqtt.rs (hunk 5 above)
Comment (satta) [reply rust/src/mqtt/mqtt.rs:241]: This is the default protocol number that is assumed and gets overwritten once we have determined the actual version from the connect handshake. In v2, I will use a Rust constant to clarify that to the reader of the code.

File: src/detect-mqtt-protocol-version.c (hunk 4 above)
Comment (satta) [reply src/detect-mqtt-protocol-version.c:118]: Done in v2, to be opened soon.

File: rust/src/mqtt/mqtt.rs (hunk 6 above)
Comment (catenacyber) [reply rust/src/mqtt/mqtt.rs:245]: `ALPROTO_UNKNOWN` means we do not have enough data to tell if it is MQTT or not
`ALPROTO_FAILED` means it is definitely not MQTT

File: rust/src/mqtt/mqtt.rs (hunk 7 above)
Comment (catenacyber) [reply rust/src/mqtt/mqtt.rs:413]: @victorjulien any thoughts here ? Should we name it `SC_DEFAULT_PROTO_MAX_DEPTH` and keep value 16 ?

File: rust/src/mqtt/mqtt.rs (hunk 5 above)
Comment (catenacyber) [reply rust/src/mqtt/mqtt.rs:241]: Thanks

File: src/detect-mqtt-protocol-version.c (hunk 4 above)
Comment (catenacyber) [reply src/detect-mqtt-protocol-version.c:118]: Thanks

File: rust/src/mqtt/mqtt.rs (hunk 6 above)
Comment (satta) [reply rust/src/mqtt/mqtt.rs:245]: Got that -- will be more decisive now!

File: src/detect-mqtt-connack-sessionpresent.c (hunk 1 above)
Comment (satta) [reply src/detect-mqtt-connack-sessionpresent.c:54]: Done.

**Comment by satta** [comment]
New PR: #4934

**Pull request closed without merging**